File: 01/1389.cpp
#include <cstdio>

int n, m;
int d[110][110];

int main()
{
	scanf("%d %d", &n, &m);
	for(int i=0; i<n; i++)
	{
		for(int j=0; j<n; j++)
			d[i][j] = 999;
		d[i][i] = 0;
	}
	for(int i=0; i<m; i++)
	{
		int a, b;
		scanf("%d %d", &a, &b);
		a--, b--;
		d[a][b] = d[b][a] = 1;
	}

	for(int k=0; k<n; k++)
		for(int i=0; i<n; i++)
			for(int j=0; j<n; j++)
				if(d[i][j] > d[i][k]+d[k][j])
					d[i][j] = d[i][k]+d[k][j];

	int min=999, idx=-1;
	for(int i=0; i<n; i++)
	{
		int sum = 0;
		for(int j=0; j<n; j++)
			sum += d[i][j];
		if(sum < min)
		{
			min = sum;
			idx = i;
		}
	}
	
	printf("%d\n", idx+1);
	return 0;
}

File: 01/1389.rs
const MAX_N: usize = 100;

fn main() {
    let (n, connectivity) = get_input();
    let answer = get_answer(n, connectivity);

    println!("{}", answer + 1);
}

fn get_input() -> (usize, Vec<(usize, usize)>) {
    let mut input = String::new();
    std::io::stdin().read_line(&mut input).unwrap();
    let mut iter = input.split_whitespace().flat_map(str::parse::<usize>);
    let mut get = || iter.next().unwrap();

    let (n, m) = (get(), get());
    let mut connectivity = vec![(0, 0); m];

    for i in 0..m {
        let mut input = String::new();
        std::io::stdin().read_line(&mut input).unwrap();
        let mut iter = input.split_whitespace().flat_map(str::parse::<usize>);
        let mut get = || iter.next().unwrap();

        connectivity[i] = (get() - 1, get() - 1);
    }

    (n, connectivity)
}

fn get_answer(n: usize, connectivity: Vec<(usize, usize)>) -> usize {
    let mut path = [MAX_N; MAX_N * MAX_N];

    let idx = |x, y| x * n + y;

    for i in 0..n {
        path[idx(i, i)] = 0;
    }
    for (a, b) in connectivity {
        path[idx(a, b)] = 1;
        path[idx(b, a)] = 1;
    }
    for k in 0..n {
        for i in 0..n {
            for j in 0..n {
                let new_path = path[idx(i, k)] + path[idx(k, j)];
                path[idx(i, j)] = path[idx(i, j)].min(new_path);
            }
        }
    }

    let min_elem = (0..n)
        .map(|x| &path[x * n..(x + 1) * n])
        .map(|s| s.iter().sum::<usize>())
        .enumerate()
        .min_by(|lhs, rhs| lhs.1.cmp(&rhs.1).then(lhs.0.cmp(&rhs.0)));

    min_elem.unwrap().0
}
